### PR TITLE
Migration to Vite

### PR DESCRIPTION
**Pull request by flomillot**
Migrate the project from create-react-app to vite

- update env variable names, replace process.env by import.meta.env
- add babel presets for tests
- remove %PUBLIC_URL%
- add jest config
- change extension file from .js into .jsx when returning react elements
- use svgr plugin for SVG and update imports
- use a plugin to fix react-virtualized bad import and update the version to works with the plugin
- update npm scripts
- replace all require by import and rename LocalizedCountries into useLocalizedCountries, also clean CountrySelect
- clean some imports (remove .js of paths)
- clean a bit utils.js and make it easier to debug on fetchEnv calls
- clean some random code for eslint since we now have some extra settings imported from commons-ui
- remove homepage from package.json (used by CRA) and set the base path via vite config.

Notes : as vite does not support dynamic imports on node_modules https://github.com/vitejs/vite/issues/14102, we have to imports all the needed language and use a if condition. Can we make it better ?

### CODE REVIEW

**Comment by EstherDarkish** [comment]
I have some console errors in the spreadsheet, like these :

- Warning: validateDOMNesting(...): <th> cannot appear as a child of <thead>

- Warning: React does not recognize the `textColor` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `textcolor` instead. If you accidentally passed it from a parent component, remove it from the DOM element.

- Warning: Received `false` for a non-boolean attribute `indicator`.

I think we should merge this PR and fix those messages later.

**Review by EstherDarkish — APPROVED** [review]
Code OK, Tests OK.
I think we should merge this and fix the small issues later. This is a big PR and the more we wait, the harder it will be to merge it.

File: index.html [hunk 1]
```diff
@@ -1,21 +1,22 @@
...
+        <base href="/" />
         <meta charset="utf-8" />
         <meta name="viewport" content="width=device-width, initial-scale=1.0" />
         <meta name="google" content="notranslate" />
-        <link rel="icon" href="favicon.ico" />
+        <link rel="icon" href="/favicon.ico" />
         <title>GridStudy</title>
     </head>
     <body class="singlestretch-parent singlestretch-child">
-        <noscript>You need to enable JavaScript to run this app.</noscript>
```
Comment (Tristan-WorkGH) [inline index.html:18]: keep this line to have a message when the navigator has scripts disable

File: index.html [hunk 2]
```diff
@@ -1,21 +1,22 @@
 <!--
-    Copyright (c) 2020, RTE (http://www.rte-france.com)
+    Copyright (c) 2024, RTE (http://www.rte-france.com)
     This Source Code Form is subject to the terms of the Mozilla Public
     License, v. 2.0. If a copy of the MPL was not distributed with this
     file, You can obtain one at http://mozilla.org/MPL/2.0/.
 -->
+
```
Comment (Tristan-WorkGH) [inline index.html:7]:
```suggestion
```

File: index.html [hunk 3]
```diff
@@ -1,21 +1,22 @@
...
 <!DOCTYPE html>
 <html class="singlestretch-parent" lang="en" translate="no">
     <head>
-        <base href="%PUBLIC_URL%/" />
+        <base href="/" />
```
Comment (Tristan-WorkGH) [inline index.html:11]: would it cause problem when use with a sub-path like on azure demo ?  like `demo.gridsuite.com/gridstudy/`

File: package.json [hunk 4]
```diff
@@ -53,26 +54,30 @@
...
     "scripts": {
-        "start": "react-scripts start",
-        "build": "react-scripts build",
-        "test": "react-scripts test --watchAll=false --transformIgnorePatterns \"node_modules/(?!@gridsuite/commons-ui)/\"",
-        "eject": "react-scripts eject"
+        "start": "vite",
+        "build": "tsc && vite build",
+        "serve": "vite preview",
+        "test": "jest",
+        "lint": "eslint . --ext js,mjs,jsx,ts,mts,tsx --max-warnings 0"
```
Comment (Tristan-WorkGH) [inline package.json:71]: `./src` to not scan `node_modules`?
```suggestion
        "lint": "eslint ./src --ext js,mjs,jsx,ts,mts,tsx --max-warnings 0"
```

File: src/services/case.ts [hunk 5]
```diff
@@ -1,7 +1,7 @@
 import { UUID } from 'crypto';
 import { backendFetchJson } from './utils';
 
-const PREFIX_CASE_QUERIES = `${process.env.REACT_APP_API_GATEWAY}/case`;
+const PREFIX_CASE_QUERIES = import.meta.env.VITE_API_GATEWAY + '/case';
```
Comment (Tristan-WorkGH) [inline src/services/case.ts:4]: why not keeping template form? not supported by vite/rollup?
```suggestion
const PREFIX_CASE_QUERIES = `${import.meta.env.VITE_API_GATEWAY}/case`;
```

File: src/services/utils.js [hunk 6]
```diff
@@ -135,16 +135,24 @@ const downloadFile = (blob, filename, type) => {
...
-        .then((env) =>
-            fetch(env.appsMetadataServerUrl + '/authentication.json')
-        )
-        .then((res) => res.json())
+        .then((env) => {
+            return fetch(env.appsMetadataServerUrl + '/authentication.json');
+        })
+        .then((res) => {
+            return res.json();
+        })
```
Comment (Tristan-WorkGH) [inline src/services/utils.js:155]: you can restore original shorter form if you return with only one statement
```suggestion
        .then((env) => fetch(env.appsMetadataServerUrl + '/authentication.json'))
        .then((res) => res.json())
```

File: src/services/utils.js [hunk 7]
```diff
@@ -174,7 +182,9 @@ export function fetchAppsAndUrls() {
 export function fetchVersion() {
     console.info(`Fetching global metadata...`);
     return fetchEnv()
-        .then((env) => fetch(env.appsMetadataServerUrl + '/version.json'))
+        .then((env) => {
+            return fetch(env.appsMetadataServerUrl + '/version.json');
+        })
```
Comment (Tristan-WorkGH) [inline src/services/utils.js:187]: same
```suggestion
        .then((env) => fetch(env.appsMetadataServerUrl + '/version.json'))
```

File: src/services/utils.js [hunk 8]
```diff
@@ -216,9 +226,7 @@ export function getUrlWithToken(baseUrl) {
...
 export function fetchMapBoxToken() {
     console.info(`Fetching MapBoxToken...`);
-    return fetch('env.json')
-        .then((res) => res.json())
-        .then((res) => {
-            return res.mapBoxToken;
-        });
+    return fetchEnv().then((res) => {
+        return res.mapBoxToken;
+    });
```
Comment (Tristan-WorkGH) [inline src/services/utils.js:231]: same
```suggestion
    return fetchEnv().then((res) => res.mapBoxToken);
```

**Review by Tristan-WorkGH — CHANGES_REQUESTED** [review]

File: index.html (hunk 1 above)
Comment (flomillot) [reply index.html:18]: Done

File: index.html (hunk 2 above)
Comment (flomillot) [reply index.html:7]: Alright

File: package.json (hunk 4 above)
Comment (flomillot) [reply package.json:71]: Nope and some other dot starting files

File: src/services/case.ts (hunk 5 above)
Comment (flomillot) [reply src/services/case.ts:4]: I prefer this form

File: index.html (hunk 3 above)
Comment (flomillot) [reply index.html:10]: Forgot to change a sed in the Dockerfile, thanks

File: src/vite-env.d.ts [hunk 9]
```diff
@@ -0,0 +1,2 @@
+/// <reference types="vite-plugin-svgr/client" />
+/// <reference types="vite/client" />
```
Comment (EstherDarkish) [inline src/vite-env.d.ts:2]: What is the purpose of this file ?

Comment (flomillot) [reply src/vite-env.d.ts:2]: For TypeScript.
Was already existing under an other name.
https://github.com/gridsuite/gridstudy-app/pull/1919/files#diff-433083e9966b3a808a0b2dce3e47b0b74152c61cb7d5b6a4ad8aad2592dacd81

File: index.html [hunk 10]
```diff
@@ -1,21 +1,22 @@
...
         <meta name="google" content="notranslate" />
-        <link rel="icon" href="favicon.ico" />
+        <link rel="icon" href="/favicon.ico" />
         <title>GridStudy</title>
     </head>
     <body class="singlestretch-parent singlestretch-child">
-        <noscript>You need to enable JavaScript to run this app.</noscript>
         <div id="root" class="singlestretch-parent singlestretch-child"></div>
+        <noscript>You need to enable JavaScript to run this app.</noscript>
+        <script type="module" src="/src/index.jsx"></script>
```
Comment (Tristan-WorkGH) [inline index.html:20]: isn't the script injected by vite during build?

File: src/components/report-viewer/report-tree-view-context.js [hunk 11]
```diff
@@ -11,7 +11,7 @@ import { createContext } from 'react';
 
 const ReportTreeViewContext = createContext({});
 
-if (process.env.NODE_ENV !== 'production') {
+if (!import.meta.env.PROD) {
```
Comment (Tristan-WorkGH) [inline src/components/report-viewer/report-tree-view-context.js:14]:
```suggestion
if (import.meta.env.DEV) {
```

**Review by Tristan-WorkGH — APPROVED** [review]

File: index.html (hunk 10 above)
Comment (flomillot) [reply index.html:20]: Sorry I didn't understand, which line ?

File: src/components/report-viewer/report-tree-view-context.js (hunk 11 above)
Comment (flomillot) [reply src/components/report-viewer/report-tree-view-context.js:14]: I missed it, I change it, thanks

**Review by etiennehomer — APPROVED** [review]

**Pull request merged**